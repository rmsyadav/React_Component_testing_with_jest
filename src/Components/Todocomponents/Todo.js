import { useReducer, useState } from "react";
import todosReducer from "../../Reducer/Reducer";
import {addTodoActionCreator,removeTodoActionCreator} from '../../Actions/Actions';
import "../../Stylesheet/Todo.css";
import Footer from "../CommonComponents/Footer";

const Todo = ()=>{
  const date = new Date('December 17, 2023 03:24:00');;
  const todayDate = date.toString();
  const [todoText,setTodoText] = useState("");
  const [todosList,dispatch] = useReducer(todosReducer);
  const [todoWaring,setTodoWaring] = useState(false);

  const AddTodoText = (event)=>{ 
   setTodoText(event.target.value);
  }

  const sleep = async (milliseconds) => {
    await new Promise(resolve => {
        return setTimeout(()=>{
            setTodoWaring(false);
            resolve();
        }, milliseconds)
    });
  };

  const addNewTodo = ()=>{
    if(todoText)
    {   setTodoWaring(false);
        dispatch(addTodoActionCreator(todoText));
    }else{
        setTodoWaring(true);
        sleep(2000);
    }
    
  }

  const deleteTodoFromTodoList = (index)=>{
    dispatch(removeTodoActionCreator(index));

  }
   return(
       <>
        <section className="container gradient-form rounded-3" style={{backgroundImage: "url('https://mdbcdn.b-cdn.net/img/Photos/new-templates/search-box/img4.webp')"}}> 
         <div className="row d-flex justify-content-center align-items-center">
            <div className="col-xl-8" style={{marginTop:"30px"}}>
                <div className="card text-center">
                    <div className="card-header bg-success text-dark fw-bold">
                        Todo App
                    </div>
                    <div className="card-body" style={{minHeight:"400px"}}>
                        {todoWaring && <div className="alert alert-warning card-title" data-testid="todoError">
                        <strong>Warning!</strong> Please type something in TodoBox field.
                        </div>}
                        <div className="input-group card-title mb-3">
                            <input type="text" className="form-control" placeholder="Add your new todo" value={todoText} onChange={AddTodoText}/>
                            <button className="btn btn-outline-secondary" type="button" id="button-addon2" onClick={addNewTodo}><i className="bi bi-plus-circle-fill" style={{color: "blue"}}></i></button>
                        </div>
                        <div className="card-text">
                        <ul className="list-group list-group-numbered">
                            {todosList && todosList.map((value,index)=>{
                                return(
                                    <li key={index} data-testid="list-group-item" className="list-group-item d-flex justify-content-between align-items-center">
                                        {value}
                                        <span><i className="bi bi-trash hoverEffect" data-testid="deleteTodo" style={{color: "red"}} onClick={()=>{deleteTodoFromTodoList(index)}}></i></span>
                                    </li>
                                    )
                            })}
                        </ul>
                        </div>    
                    </div>
                    <div className="card-footer text-light bg-warning">
                        {todayDate}
                    </div>
            </div>
            </div>
         </div> 
        </section> 
        <Footer></Footer>
      </>   
   ) 
}

export default Todo;