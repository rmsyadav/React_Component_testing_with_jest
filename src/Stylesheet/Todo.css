.container {
    width: 50%;
    padding-right: 0 !important;
    padding-left: 0 !important;
    margin-top: 30px;
    margin-right: auto;
    margin-left: auto;
}
.card-text{
    width: 80%;
    margin-right: auto;
    margin-left: auto;
}

.list-group-item:hover{
    background-color: chartreuse;
}
