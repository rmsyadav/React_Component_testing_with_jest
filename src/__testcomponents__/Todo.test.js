import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event'
import Todo from '../Components/Todocomponents/Todo';

// App component testing for rendering is properly or not

describe("App component testing:- ",()=>{

  it('Todo component snapshot testing ',()=>{

    const todo = renderer.create(<Todo></Todo>).toJSON();
    expect(todo).toMatchSnapshot();
  })

  test('renders the todo component', () => {

    render(<Todo></Todo>);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeInTheDocument();
    const textBoxElement = screen.getByRole('textbox');
    expect(textBoxElement).toBeInTheDocument();
    
  });

  test('Perform the user action',async()=>{
    const user = userEvent.setup();
    render(<Todo></Todo>);
    const textBoxElement = screen.getByRole('textbox');
    await user.type(textBoxElement,'this is todo 1')
    expect(screen.getByRole('textbox')).toHaveValue('this is todo 1')
    const buttonElement = screen.getByRole('button');
    await user.click(buttonElement);
    const displayedTasks = await screen.findAllByTestId("list-group-item");
    expect(displayedTasks).toHaveLength(1);
    expect(screen.getByText('this is todo 1')).toBeInTheDocument();

    
  })
  

})

