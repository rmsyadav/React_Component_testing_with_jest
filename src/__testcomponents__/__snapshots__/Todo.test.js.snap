// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component testing:-  Todo component snapshot testing  1`] = `
Array [
  <section
    className="container gradient-form rounded-3"
    style={
      Object {
        "backgroundImage": "url('https://mdbcdn.b-cdn.net/img/Photos/new-templates/search-box/img4.webp')",
      }
    }
  >
    <div
      className="row d-flex justify-content-center align-items-center"
    >
      <div
        className="col-xl-8"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <div
          className="card text-center"
        >
          <div
            className="card-header bg-success text-dark fw-bold"
          >
            Todo App
          </div>
          <div
            className="card-body"
            style={
              Object {
                "minHeight": "400px",
              }
            }
          >
            <div
              className="input-group card-title mb-3"
            >
              <input
                className="form-control"
                onChange={[Function]}
                placeholder="Add your new todo"
                type="text"
                value=""
              />
              <button
                className="btn btn-outline-secondary"
                id="button-addon2"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="bi bi-plus-circle-fill"
                  style={
                    Object {
                      "color": "blue",
                    }
                  }
                />
              </button>
            </div>
            <div
              className="card-text"
            >
              <ul
                className="list-group list-group-numbered"
              />
            </div>
          </div>
          <div
            className="card-footer text-light bg-warning"
          >
            Sun Dec 17 2023 03:24:00 GMT+0530 (India Standard Time)
          </div>
        </div>
      </div>
    </div>
  </section>,
  <footer
    className="container-fluid footer-img rounded-2"
  >
    <div
      className="text-center footer-fontsize-mobile"
    >
      © 2023 Copyright:
      <span
        className="text-reset fw-bold"
      >
         Learning App
      </span>
    </div>
  </footer>,
]
`;
